File: python/examples/cdp-crewai-chatbot/chatbot.py
import os
import sys
from dotenv import load_dotenv
from langchain_openai import ChatOpenAI
from crewai import Agent, Crew, Task
from cdp_crewai.agent_toolkits import CdpToolkit
from cdp_crewai.utils import CdpAgentkitWrapper

# Configure a file to persist the agent's CDP MPC Wallet Data
wallet_data_file = "wallet_data.txt"

load_dotenv()

def validate_environment():
    """Validate that required environment variables are set."""
    required_vars = ["OPENAI_API_KEY", "CDP_API_KEY_NAME", "CDP_API_KEY_PRIVATE_KEY"]
    missing_vars = [var for var in required_vars if not os.getenv(var)]
    
    if missing_vars:
        print("Error: Required environment variables are not set")
        for var in missing_vars:
            print(f"{var}=your_{var.lower()}_here")
        sys.exit(1)
    
    if not os.getenv("NETWORK_ID"):
        print("Warning: NETWORK_ID not set, defaulting to base-sepolia testnet")

def initialize_web3_agent():
    """Initialize the Web3 Agent with CDP tools."""
    print("Initializing Web3 Agent...")

    wallet_data = None

    if os.path.exists(wallet_data_file):
        with open(wallet_data_file) as f:
            wallet_data = f.read()

    # Configure CDP Agentkit Langchain Extension.
    values = {}
    if wallet_data is not None:
        # If there is a persisted agentic wallet, load it and pass to the CDP Agentkit Wrapper.
        values = {"cdp_wallet_data": wallet_data}
    
    # Initialize CDP wrapper and toolkit
    cdp_wrapper = CdpAgentkitWrapper(**values)
    toolkit = CdpToolkit.from_cdp_agentkit_wrapper(cdp_wrapper)
    
    # Create the Web3 Agent with tracing disabled
    web3_agent = Agent(
        role='Web3 Agent',
        goal='Help users interact with blockchain',
        backstory='I am an expert in blockchain interactions and Web3 technologies',
        tools=toolkit.get_tools(),
        verbose=False,
        llm=ChatOpenAI(model_name="gpt-4o"),

    )
    
    return web3_agent

def create_task(prompt: str, agent) -> Task:
    """Create a task for the agent."""
    return Task(
        description=prompt,
        agent=agent,
        expected_output="Successfully completed Web3 interaction or provided relevant information"
    )

def main():
    validate_environment()
    web3_agent = initialize_web3_agent()
    print("Agent ready! Type 'exit' to end the chat\n")
    
    while True:
        prompt = input("\nPrompt: ")
        if prompt.lower() == 'exit':
            break
            
        try:
            task = create_task(prompt, web3_agent)
            # Disable tracing and configure crew with minimal output
            os.environ["LANGCHAIN_TRACING"] = "false"
            os.environ["OPENAI_TRACING"] = "false"
            
            crew = Crew(
                agents=[web3_agent],
                tasks=[task],
                verbose=False,
                process_inputs=False
            )
            result = crew.kickoff()
            print("\n-------------------")
            print(result)
            print("-------------------")
        except Exception as e:
            print(f"Error: {str(e)}")
            print("Please try your request again")

if __name__ == "__main__":
    main() 

File: typescript/examples/vercel-ai-sdk-cdp-chatbot/chatbot.ts
import { AgentKit, cdpApiActionProvider, erc721ActionProvider, pythActionProvider, walletActionProvider, CdpWalletProvider } from "@coinbase/agentkit";
import { getVercelAITools } from "@coinbase/agentkit-vercel-ai-sdk";
import { openai } from "@ai-sdk/openai";
import { generateId, Message, streamText, ToolSet } from "ai";
import * as dotenv from "dotenv";
import * as readline from "readline";
import * as fs from "fs";

dotenv.config();

/**
 * Validates that required environment variables are set
 *
 * @throws {Error} - If required environment variables are missing
 * @returns {void}
 */
function validateEnvironment(): void {
  const missingVars: string[] = [];

  // Check required variables
  const requiredVars = ["OPENAI_API_KEY", "CDP_API_KEY_NAME", "CDP_API_KEY_PRIVATE_KEY"];
  requiredVars.forEach(varName => {
    if (!process.env[varName]) {
      missingVars.push(varName);
    }
  });

  // Exit if any required variables are missing
  if (missingVars.length > 0) {
    console.error("Error: Required environment variables are not set");
    missingVars.forEach(varName => {
      console.error(`${varName}=your_${varName.toLowerCase()}_here`);
    });
    process.exit(1);
  }

  // Warn about optional NETWORK_ID
  if (!process.env.NETWORK_ID) {
    console.warn("Warning: NETWORK_ID not set, defaulting to base-sepolia testnet");
  }
}

// Add this right after imports and before any other code
validateEnvironment();

// Configure a file to persist the agent's CDP MPC Wallet Data
const WALLET_DATA_FILE = "wallet_data.txt";

const system = `You are a helpful agent that can interact onchain using the Coinbase Developer Platform AgentKit. You are
empowered to interact onchain using your tools. If you ever need funds, you can request them from the
faucet if you are on network ID 'base-sepolia'. If not, you can provide your wallet details and request
funds from the user. Before executing your first action, get the wallet details to see what network
you're on. If there is a 5XX (internal) HTTP error code, ask the user to try again later. If someone
asks you to do something you can't do with your currently available tools, you must say so, and
encourage them to implement it themselves using the CDP SDK + Agentkit, recommend they go to
docs.cdp.coinbase.com for more information. Be concise and helpful with your responses. Refrain from
restating your tools' descriptions unless it is explicitly requested.`

/**
 * Initialize the agent with CDP Agentkit and Vercel AI SDK tools
 *
 * @returns Object containing initialized tools
 * @throws Error if initialization fails
 */
async function initializeAgent() {
  try {

    let walletDataStr: string | null = null;

    // Read existing wallet data if available
    if (fs.existsSync(WALLET_DATA_FILE)) {
      try {
        walletDataStr = fs.readFileSync(WALLET_DATA_FILE, "utf8");
      } catch (error) {
        console.error("Error reading wallet data:", error);
        // Continue without wallet data
      }
    }

    // Configure CDP Wallet Provider
    const config = {
      apiKeyName: process.env.CDP_API_KEY_NAME,
      apiKeyPrivateKey: process.env.CDP_API_KEY_PRIVATE_KEY?.replace(/\\n/g, "\n"),
      cdpWalletData: walletDataStr || undefined,
      networkId: process.env.NETWORK_ID || "base-sepolia",
    };

    const walletProvider = await CdpWalletProvider.configureWithWallet(config);

    // Initialize action providers
    const cdp = cdpApiActionProvider({
      apiKeyName: process.env.CDP_API_KEY_NAME,
      apiKeyPrivateKey: process.env.CDP_API_KEY_PRIVATE_KEY,
    });
    const erc721 = erc721ActionProvider();
    const pyth = pythActionProvider();
    const wallet = walletActionProvider();

    const agentKit = await AgentKit.from({ 
      walletProvider,
      actionProviders: [cdp, erc721, pyth, wallet],
    });

    // Log available actions
    const actions = agentKit.getActions();
    for (const action of actions) {
      console.log(`Available action: ${action.name}`);
    }

    const tools = getVercelAITools(agentKit);
    return { tools };
  } catch (error) {
    console.error("Failed to initialize agent:", error);
    throw error;
  }
}

/**
 * Run the chatbot in interactive mode
 *
 * @param tools - Record of Vercel AI SDK tools from AgentKit
 * @returns Promise that resolves when chat session ends
 */
async function runChatMode(tools: ToolSet) {
  console.log("Starting chat mode... Type 'exit' to end.");

  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  const question = (prompt: string): Promise<string> =>
    new Promise(resolve => rl.question(prompt, resolve));

  const messages: Message[] = [];
  let running = true;

  try {
    while (running) {
      const userInput = await question("\nPrompt: ");

      if (userInput.toLowerCase() === "exit") {
        running = false;
        continue;
      }

      messages.push({ id: generateId(), role: "user", content: userInput });

      const stream = streamText({
        model: openai("gpt-4-turbo-preview"),
        messages,
        tools,
        system,
        maxSteps: 10,
      });

      let assistantMessage = "";
      for await (const chunk of stream.textStream) {
        process.stdout.write(chunk);
        assistantMessage += chunk;
      }
      console.log("\n-------------------");

      messages.push({ id: generateId(), role: "assistant", content: assistantMessage });
    }
  } catch (error) {
    console.error("Error:", error);
  } finally {
    rl.close();
  }
}

/**
 * Main entry point for the chatbot application
 * Initializes the agent and starts chat mode
 *
 * @throws Error if initialization or chat mode fails
 */
async function main() {
  try {
    const { tools } = await initializeAgent();
    await runChatMode(tools);
  } catch (error) {
    console.error("Error:", error);
    process.exit(1);
  }
}

if (require.main === module) {
  console.log("Starting Agent...");
  main().catch(error => {
    console.error("Fatal error:", error);
    process.exit(1);
  });
}
